service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/AssignManagedIdentityAzureVmStep.java

```diff
@@ -0,0 +1,95 @@
+import com.azure.core.management.profile.AzureProfile;
+import com.azure.resourcemanager.compute.ComputeManager;
+import com.azure.resourcemanager.msi.MsiManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AssignManagedIdentityAzureVmStep implements Step {
+
+  private static final Logger logger =
+      LoggerFactory.getLogger(AssignManagedIdentityAzureVmStep.class);
+  private final AzureConfiguration azureConfig;
+  private final CrlService crlService;
+  private final SamService samService;
+  private final AuthenticatedUserRequest userRequest;
+  private final ControlledAzureVmResource resource;
+
+  public AssignManagedIdentityAzureVmStep(
+      AzureConfiguration azureConfig,
+      CrlService crlService,
+      SamService samService,
+      AuthenticatedUserRequest userRequest,
+      ControlledAzureVmResource resource) {
+    this.azureConfig = azureConfig;
+    this.crlService = crlService;
+    this.samService = samService;
+    this.userRequest = userRequest;
+    this.resource = resource;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    FlightMap inputMap = context.getInputParameters();
+    FlightUtils.validateRequiredEntries(inputMap, ControlledResourceKeys.CREATION_PARAMETERS);
+
+    final AzureCloudContext azureCloudContext =
+        context
+            .getWorkingMap()
+            .get(ControlledResourceKeys.AZURE_CLOUD_CONTEXT, AzureCloudContext.class);
+
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+    AzureProfile profile =
+        new AzureProfile(
+            azureCloudContext.getAzureTenantId(),
+            azureCloudContext.getAzureSubscriptionId(),
+            AzureEnvironment.AZURE);
+    MsiManager msiManager =
+        MsiManager.configure()
+            .authenticate(crlService.getManagedAppCredentials(azureConfig), profile);
+
+    String petManagedIdentityId =
+        SamRethrow.onInterrupted(
+            () ->
+                samService.getOrCreateUserManagedIdentity(
+                    userRequest,
+                    azureCloudContext.getAzureSubscriptionId(),
+                    azureCloudContext.getAzureTenantId(),
+                    azureCloudContext.getAzureResourceGroupId()),
+            "getPetManagedIdentity");
+
+    return AzureVmHelper.assignPetManagedIdentityToVm(
+        azureCloudContext, computeManager, msiManager, resource.getVmName(), petManagedIdentityId);
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    final AzureCloudContext azureCloudContext =
+        context
+            .getWorkingMap()
+            .get(ControlledResourceKeys.AZURE_CLOUD_CONTEXT, AzureCloudContext.class);
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+
+    String petManagedIdentityId =
```

Comment:
It feels slightly strange to be calling Sam in the `undoStep`. Undo may have been called if the Sam call fails, in which case we'd call it (and fail) again.

Maybe undo should just call `AzureVmHelper.removeAllUserAssignedManagedIdentitiesFromVm`? (Since we only ever assign 1 anyway.)

Reply:
Yes, a bit of overcomplication here to avoid removing all MIs. :)